### ISSUE TITLE
Temporary basal history disappears (NS 15.0.2 + AAPS 3.1.0.2)

### ISSUE THREAD

**Issue by ireneusz-ptak** [nightscout/cgm-remote-monitor#8147]
**Describe the bug**
Not sure if it's a NS or AAPS problem. It seems that (some) temporary basal changes disappear from NS when new data arrives.

You may see it on screenshots - between 20:15 and 20:30 a 0% basal is shown, when the percentage changes to 200% the previous entry disappears. What's worse is that it doesn't always happen and I'm not yet sure what is the key here (just switched to NS 15 and AAPS 3.2).

**Expected behavior**
Proper temporary basal displayed.

**Screenshots**
![Screenshot from 2023-10-27 20-29-46](https://github.com/nightscout/cgm-remote-monitor/assets/31506973/2e95751e-a865-4bea-a7e8-8f1dbc5b3d2f)
![Screenshot from 2023-10-27 20-32-46](https://github.com/nightscout/cgm-remote-monitor/assets/31506973/25b1e5d9-bb5b-40fb-9297-f0a78f87f1c7)

**Your setup information**
NS 15.0.2 (docker)
AAPS 3.1.0.2 (NSClientV3 + websockets)

**Comment by ireneusz-ptak** [comment]
@bewest - I haven't seen anyone else reporting it, might be introduced in 15.0.2?

**Comment by ireneusz-ptak** [comment]
I've just reverted to 15.0.1 and the basal history is back.

**Comment by olorinmaia** [comment]
Experiencing same issue. AAPS 3.2 and NS 15.0.2. Following.

**Comment by olorinmaia** [comment]
Some further testing show that basalhistory with NSClientV1 sync is OK but NSClientV3 it's not 100% right back in history. 

@bewest & @MilosKozak will this be fixed when NS15 internally support v3api or is it something else causing this?

AAPS 3.2.0.1-dev
Build: ddc643b83a-2023.10.31
Nightscout-versjon: master 15.0.2

Screenshot NS:
![image](https://github.com/nightscout/cgm-remote-monitor/assets/114103483/922124d8-ca55-460c-9c03-04382230c58f)

AAPS 3.2.0.1-dev:
![image](https://github.com/nightscout/cgm-remote-monitor/assets/114103483/dcc9e16d-7c5f-44b3-8152-3ca36e7b9f6e)
Logs
[AndroidAPS_LOG_1698791493172.log.zip](https://github.com/nightscout/cgm-remote-monitor/files/13221802/AndroidAPS_LOG_1698791493172.log.zip)

**Comment by jimk9** [comment]
I have similar issue. Pics of graphs in aaps and navis attached as well as pics showing treatment tab. AAPS has 0 temp basal. Navis shows basal for that time.
3.2.0.1
15.0.2
Uploads only from AAPS using V3
![nightscout](https://github.com/nightscout/cgm-remote-monitor/assets/52563974/2bc42854-d824-432a-b262-6708774d16e2)
![aaps](https://github.com/nightscout/cgm-remote-monitor/assets/52563974/de8f9ffc-0c48-4d51-8c1c-984d0a2d7459)
![treatments-09-11](https://github.com/nightscout/cgm-remote-monitor/assets/52563974/e5289270-dade-4ed1-9ad6-e59371292b3d)
![treatments-13-15](https://github.com/nightscout/cgm-remote-monitor/assets/52563974/5d66a382-842f-4f3f-b9a3-debba7a089f9)

**Comment by jimk9** [comment]
I think #7771 is the same issue.

Will look to see if there are any more issues that could be linked

**Comment by ireneusz-ptak** [comment]
> Can you check to see if your treatment information in reports in nightscout agree with AAPS

It looks that everything is the same (I cannot check the exact moment from the screenshots, because I broke the DB while trying to update the Mongo version, but I was able to compare two other days).

**Comment by ireneusz-ptak** [comment]
> I think #7771 is the same issue.
> 
> Will look to see if there are any more issues that could be linked

Not sure about that - the problem disappears after switching back to NSClientV1 (from NSClientV3) in AAPS, so it's probably related to V3 API.

**Comment by jimk9** [comment]
Yep I understand that and to be honest I'm not sure either. What confuses me is that nightscout has all the correct information in the reports which suggests the communication between nightscout and aaps was complete and the database is accurate. Viewing the data in treatments report or in the daily report it all looks correct. Only bit that is wrong is the live basal graph. The good thing for me is the reports are accurate so I'm going to stay with v3 at the moment. Not sure anyone is looking at this issue

**Comment by sulkaharo** [comment]
This sounds like AAPS is doing something different with the 0 basal marking from what it's stored before. I think @MilosKozak might be the only NS dev who also uses the new AAPS and is thus able to debug what's happening.

**Comment by MilosKozak** [comment]
It's rendering issue

**Comment by yuvsketch** [comment]
Same problem here...
Any solution?

![Screenshot 2023-11-15 at 10 02 14](https://github.com/nightscout/cgm-remote-monitor/assets/75831100/319476f9-6f87-4687-b2cd-76f75739b8b1)
![Uploading WhatsApp Image 2023-11-15 at 10.02.10.jpeg…]()

**Comment by roco5** [comment]
I have  the same problem with V1 of API, and Nightscout 15.0.2 and AAPSClient 3.2.04.  Temp basal always shows 0.

(V3 API presents other issues with Nightscout 15.0.2 and AAPSClient 3.2.04, like not being able to update profiles and not showing COB.)

Is there a solution that works at least as well as NS 14 and AAPS 3.1 ?


![Screenshot_20240429-074400](https://github.com/nightscout/cgm-remote-monitor/assets/6946920/e94b3a94-8912-44e7-8795-413e0ae4cbc9)
![Screenshot_20240429-074032](https://github.com/nightscout/cgm-remote-monitor/assets/6946920/e97a1ddb-1677-4959-9047-112e0c4836ec)

**Comment by psonnera** [comment]
I've been trying to troubleshoot this for some time now. You'll see basal is rendered correctly when you restart Nightscout, so it's not a rendering issue. It's a cache issue: if you disable the cache in lib/api3/storage/mongoCachedCollection/index.js as 
`function cacheSupported () {
  return false;
}`
The issue is not present. Now troubleshooting Nightscout cache is out of my abilities... so I'll leave my instance like that.

**Comment by MilosKozak** [comment]
I can confirm this will resolve the issue

**Comment by olorinmaia** [comment]
With the help of admin of ns.10be.de we have added the fix proposed by @psonnera. I have tested it and it seem to be fixing the main issue in this thread. If users of 10be.de wants this change to take effect they need to hit redeploy on the server admin page at 10be.de

**Comment by ireneusz-ptak** [comment]
Do we know what the impact on performance is? What do you think about disabling it in the repo itself? I'm keen on using the workaround but I'm using the official docker image.

**Comment by ireneusz-ptak** [comment]
@sulkaharo ?

**Comment by psonnera** [comment]
@olorinmaia I don't think it's a fix: just a workaround... as said by @ireneusz-ptak I'm afraid it will affect performance, even if in my case, hosting both containers on the same VPS I don't think it's too impacting. Still, I'd like to fix it and I'm still on it.

**Comment by olorinmaia** [comment]
> [@olorinmaia](https://github.com/olorinmaia) I don't think it's a fix: just a workaround... as said by [@ireneusz-ptak](https://github.com/ireneusz-ptak) I'm afraid it will affect performance, even if in my case, hosting both containers on the same VPS I don't think it's too impacting. Still, I'd like to fix it and I'm still on it.

Yes I know. I'm not English native so bad choice of words. But I think it's worth using this workaround for time being.

**Comment by psonnera** [comment]
@olorinmaia let me continue testing, last attempt gave a crazy result: I re-enabled cache and added some console.logs to trace the data flow inside the cache... to my big surprise basal rendering was good overnight. Since the bug needs some hours to show, I'll tell you more later.

**Comment by olorinmaia** [comment]
I can see that Temp Targets get rendered in full duration regardless if they are cancelled or not with cached disabled. So it fixes one thing and breaks another. However, TT display using v3 api is not working properly in NSClientV3 with cache enabled either.

**Comment by psonnera** [comment]
Attempt # 2 try this https://github.com/nightscout/cgm-remote-monitor/commit/52450f8b48d93446ab2fc5614b9b8c3c685001bc

<img width="1910" height="598" alt="Image" src="https://github.com/user-attachments/assets/836eaf09-9e73-469c-af3f-846650749e85" />

There are still pieces missing but few. For me it's a good workaround: not seeing basal rates in Nightscout makes visual evaluation a pain. Nightscout reports are also affected by the glitches (probably events timestamps linked).

Note AAPSClient is sync'ed with AAPS on basal rates.

Note: reboot exposes the real Nightscout database contents, rebooting Nightscout fixes the basal display bug, so I add it as aa reference.
